src/kernel-install/ukify.py

```diff
@@ -0,0 +1,233 @@
+#!/usr/bin/python
+# SPDX-License-Identifier: LGPL-2.1+
+
+import argparse
+import dataclasses
+import itertools
+import pathlib
+import os
+import shlex
+import subprocess
+import tempfile
+import typing
+
+efi_arches = {
+        # host_arch: efi_arch
+        'x86'    : 'ia32',
+        'x86_64' : 'x64',
+        'arm'    : 'arm',
+        'aarch64': 'aa64',
+        'riscv64': 'riscv64',
+}
+
+def guess_efi_arch():
+    arch = os.uname().machine
+    efi_arch = efi_arches[arch]
+
+    if arch == 'x86_64':
+        try:
+            size = open('/sys/firmware/efi/fw_platform_size').read().strip()
+        except FileNotFoundError:
+            pass
+        else:
+            if int(size) == 32:
+                efi_arch = 'ia32'
+
+    print(f'Host arch {arch!r}, efi arch {efi_arch!r}')
+    return efi_arch
+
+def shell_join(cmd):
+    return ' '.join(shlex.quote(str(x)) for x in cmd)
+
+@dataclasses.dataclass
+class Section:
+    name: str
+    content: str
+    offset: int
+    tmpfile: typing.Optional[typing.IO] = None
+
+    @classmethod
+    def create(cls, name, content, offset):
+        if isinstance(content, str):
+            tmp = tempfile.NamedTemporaryFile(mode='wt', prefix=f'tmp{name}')
+            tmp.write(content)
+            tmp.flush()
+            content = tmp.name
+        else:
+            tmp = None
+
+        return cls(name, content, offset, tmpfile=tmp)
+
+    @classmethod
+    def parse_arg(cls, s):
+        parts = s.split(':')
+        if len(parts) < 2 or len(parts) > 3:
+            raise ValueError(f'Cannot parse section spec: {s!r}')
+        name, content = parts[:2]
+        offset = int(parts[2], base=0) if len(parts) >= 3 else None
```

Comment:
This could destructure directly, e.g.
```suggestion
        try:
            name, content, *offset = s.split(':')
        except ValueError as e:
            raise ValueError(f'Cannot parse section spec: {s!r}') from e
        offset = int(offset[0], base=0) if offset else None
```

Reply:
The above version is more in line with the `len(parts) >= 3` check in the offset parsing line. If you want to throw an error for `len(parts) > 3`, than that's not in this version and would need to be added. I think it's a bit more readable this way, though.

Reply:
I think that an error should be thrown if there's too many parts. And then the version with `try..except` becomes noticably longer… Dunno, I think it's fine as is.

Reply:
Yeah, I wasn't all too sure about this myself, because of the later check for `len(parts) >= 3`. It is fine as is, but adding another check does allow for nicer error messages, including @medhefgo's waning
```python       
try:
    name, content, *offset = s.split(':')
except ValueError as e:
    raise ValueError(f'Cannot parse section spec (name or content missing): {s!r}') from e
if len(offset) > 1:
    raise ValueError(f'Cannot parse section spec (more than one offset given): {s!r}')
if len(name) > 8:
    print('Warning: name of PE section is too long')
offset = int(offset[0], base=0) if offset else None